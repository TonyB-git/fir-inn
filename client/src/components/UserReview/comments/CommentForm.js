import { useState } from "react";

const CommnetForm = ({handleSubmit, 
                        submitLable, 
                        hasCancelButton = false, 
                        initialText = '',
                        handleCancel
    }) => {
    
    const [text, setText] = useState(initialText);
    
    // Text area is empty, can't click submit button
    const isTextareaDisabled = text.length === 0;

    const onSubmit = event => {
        event.preventDefault()
        handleSubmit(text)
        // After submiting the text, the text area becomes empty
        setText("")
    }

    return (
        <form onSubmit={onSubmit}>
            <textarea className="comment-form-textarea" 
            value={text} 
            onChange={(e) => setText(e.target.value)}
        />
        <button className="comment-form-button" disabled={isTextareaDisabled}>{submitLable}</button>
        {hasCancelButton && (
            <button type = "button" className="comment-form-button comment-form-cancel-button"
            onClick={handleCancel}
            >
                Cancel
            </button>
        )}
        </form>
    ); 
};

export default CommnetForm;