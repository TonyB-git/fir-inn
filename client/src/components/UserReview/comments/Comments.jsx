import React, { Component }  from 'react';
import { useEffect, useState } from "react";
import { getComments as getCommentsApi, 
        createComment as createCommentApi, 
        deleteComment as deleteCommentApi,
        updateComment as updateCommentApi} from "../api";
import Commnet from "./Comment";
import CommentForm from "./CommentForm"

const Commnets = ({currentUserId}) => {

    const[backendComments, setBackedComments] = useState([]);
    
    const [activeComment, setActiveComment] = useState(null);


    const rootComments = backendComments.filter(
        (backendComments) => backendComments.parentId === null
    );

    const getReplies = commenId => {
        return backendComments.filter(backendComments => backendComments.parentId === commenId)
        .sort((a,b) => 
        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime());
    };

    const addComment = (text, parentId) => {
        console.log("addComment", text, parentId);
        createCommentApi(text, parentId).then(comment => {
            setBackedComments([comment, ...backendComments])
            // Close the reply box
            setActiveComment(null);
        });
    };

    const deleteComment = (commenId) => {
        if(window.confirm('Are you sure to remove?')){
            deleteCommentApi(commenId).then(() => {
                const updatedBackendComments = backendComments.filter(
                    backendComments => backendComments.id !== commenId);
                    setBackedComments(updatedBackendComments);
            });
        }
    };

    const updateComment = (text, commenId) => { 
        updateCommentApi(text, commenId).then(() => {
            const updatedBackendComments = backendComments.map(backendComments => {
                if(backendComments.id === commenId){
                    return {...backendComments, body: text};
                }
                return backendComments;
            });
            setBackedComments(updatedBackendComments);
            setActiveComment(null);
        });
    };


    // Fetch data 
    useEffect(() => {
        getCommentsApi().then(data => {
            setBackedComments(data);
        })
    }, [])
 
    return (
        <div className="comments">
            <h3 className="comments-title">Comments</h3>
            <div className="comment-form-title">Write comment</div>
            <CommentForm submitLable="Write" handleSubmit={addComment}/>
            <div className="comments-container">
                {rootComments.map((rootComment) => (
                    // <div key={rootComment.id}>{rootComment.body}</div>
                    <Commnet 
                        key={rootComment.id} 
                        comment={rootComment} 
                        replies={getReplies(rootComment.id)}
                        currentUserId={currentUserId}
                        delteComment={deleteComment}
                        updateComment={updateComment}
                        activeComment={activeComment}
                        setActiveComment={setActiveComment}
                        addComment={addComment}
                    />
                ))}
            </div>
        </div>
    );
};

export default Commnets;