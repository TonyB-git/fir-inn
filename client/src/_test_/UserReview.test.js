import renderer from 'react-test-renderer';
import {render, screen, cleanup} from '@testing-library/react';
import '@testing-library/jest-dom';
import Comments from '../components/UserReview/comments/Comments.jsx';
import CommentForm from '../components/UserReview/comments/CommentForm';

import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// React 16 Enzyme adapter
configure({ adapter: new Adapter() });

import { useEffect, useState } from "react";


// it('rendering UserReview', () => {

//     const component = renderer.create(
//         <Comments />
//     );

//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();

//     renderer.act(() => {
//         tree.props.addComment();
//       });
// }); 

// afterEach(() => {
//     cleanup();
// });

// test("Render UserReview", () => {
//     render(<Comments />);

//     var comEle = screen.getByTitle("comments-title");

//     expect(comEle).toBeInTheDocument();
// });


it("renders UserReview title", () => {
    const wrapper = shallow(<Comments />);
    const welcome =  <h3 className="comments-title">Comments</h3>;
    expect(wrapper.contains(welcome)).toEqual(true);

});

it("renders UserReview class", () => {
    const wrapper = shallow(<Comments />);
    const func1 = <div className="comment-form-title">Write comment</div>;
    expect(wrapper.contains(func1));
});  

it("renders UserReview function", () => {
    const wrapper = shallow(<Comments />);
    const func2 =  <div className="comments-container"></div>;
    expect(wrapper.contains(func2));
});  