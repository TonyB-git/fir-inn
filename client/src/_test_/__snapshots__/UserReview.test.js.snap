// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering UserReview 1`] = `
<div
  className="comments"
>
  <h3
    className="comments-title"
  >
    Comments
  </h3>
  <div
    className="comment-form-title"
  >
    Write comment
  </div>
  <form
    onSubmit={[Function]}
  >
    <textarea
      className="comment-form-textarea"
      onChange={[Function]}
      value=""
    />
    <button
      className="comment-form-button"
      disabled={true}
    >
      Write
    </button>
  </form>
  <div
    className="comments-container"
  />
</div>
`;
